torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
   void (*kernel)(uint32_t, void**) = nullptr;
 };
 
-std::shared_ptr<CompiledFusionFunction> FusionCompiler::getOrCompile(AnnotatedGraph & agraph) {
-  std::stringstream key;
-  key << *agraph.graph << "\n";
-  key << "device " << agraph.device << "\n";
-  for(auto & i : agraph.input_desc)
-    key << i << "\n";
-  for(auto & i : agraph.output_desc)
-    key << i << "\n";
-  std::string key_ = key.str();
+////////////////////////////////////////////////////////////////////////////////
+// FusedKernelCache
+
+// Note [Run-time shape checking code]
+// There are multiple assumptions that our codegen makes, which we can't check
+// in the fusion pass, because we don't have the shape information. Most notably,
+// that all values (post-input-chunk, and pre-output-concat) have the same shape
+// (hereinafter referred to as map size). One way to check this would be to run
+// shape propagation for every size configuration we get as an input, but that
+// requires a full graph traversal, and might incur unnecessary overhead. The code
+// below uses a few nice properties of broadcasting rules and their interactions with
+// pointwise operations, and takes a smarter approach, to quickly verify validity of
+// the kernel.
+//
+// Notation:
+//   - a.s when a is a tensor is a shorthand for a.shape.
+//   - B is a shorthand for the broadcasting/expanding function. It is used as a
+//     vararg function.
+//   - E is a shorthand for expand function.
+//   - Every pointwise operation can be equivalently rewritten as
+//     f(a, b) = f^(E(a, B(a.s, b.s)), E(b, B(a.s, b.s))),
+//     where f^ is a non-broadcasting verison of f.
+//   - A set of inputs that are used to produce a certain graph output is referred to
+//     as the output's broadcasting group (see Lemma 2. for explanation why).
+//
+// Lemma 1. Set of lists of integers (shapes) + { _|_ (bottom/error marker) }, with the
+//          operation of broadcasting (returning bottom upon shape mismatch) forms a monoid.
+//          In simpler terms: broadcasting is associative, i.e. B(a, B(b, c)) == B(B(a, b), c).
+//
+// Proof.   Satisfies all monoid laws:
+//            - Closed under broadcasting (trivial)
+//            - Empty shape is the identity element: B(a, []) == B([], a) == a
```

Comment:
By empty shape you mean a scalar, right? Because this could be confused with an empty tensor (numel == 0), where any operation on a n empty tensor returns an empty tensor